contracts/external/cw-payroll/src/contract.rs

```diff
@@ -0,0 +1,1462 @@
+            let mut stream = STREAMS
+                .may_load(storage, stream_id)?
+                .ok_or(ContractError::StreamNotFound { stream_id: id })?;
+            if stream.admin != info.sender {
+                return Err(ContractError::Unauthorized {});
+            }
+            if stream.paused {
+                return Err(ContractError::StreamAlreadyPaused {});
+            }
+            stream.paused_time = Some(env.block.time.seconds());
+            stream.paused = true;
+            save_stream(storage, id, &stream).unwrap();
+            Ok(stream)
+        };
+    let stream = pause_stream_local(id, deps.storage).unwrap();
+
+    if let Some(link_id) = stream.link_id {
+        pause_stream_local(link_id, deps.storage).unwrap();
+    }
+    let response = Response::new()
+        .add_attribute("method", "pause_stream")
+        .add_attribute("paused", stream.paused.to_string())
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("admin", info.sender)
+        .add_attribute("paused_time", stream.paused_time.unwrap().to_string());
+
+    Ok(response)
+}
+pub fn execute_remove_stream(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: StreamId,
+) -> ContractResult {
+    let stream = STREAMS
+        .may_load(deps.storage, id)?
+        .ok_or(ContractError::StreamNotFound { stream_id: id })?;
+
+    if stream.admin != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+    if let Some(link_id) = stream.link_id {
+        return Err(ContractError::LinkedStreamDeleteNotAllowed { link_id });
+    }
+    remove_stream(deps.storage, id).unwrap();
+    let response = Response::new()
+        .add_attribute("method", "remove_stream")
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("admin", info.sender)
+        .add_attribute("removed_time", env.block.time.to_string())
+        .add_message(
+            create_balance_transfer_msg(&stream.balance, None, stream.admin.clone().into())
+                .unwrap(),
+        );
+    Ok(response)
+}
+fn create_balance_transfer_msg(
+    balance: &WrappedBalance,
+    amount: Option<Uint128>,
+    recepient: String,
+) -> Result<CosmosMsg, ContractError> {
+    if let Some(native_balance) = balance.native() {
+        let msg = CosmosMsg::Bank(BankMsg::Send {
+            to_address: recepient,
+            amount: vec![Coin {
+                denom: native_balance.denom.clone(),
+                amount: amount.unwrap_or(native_balance.amount),
+            }],
+        });
+        return Ok(msg);
+    } else if let Some(cw20_balance) = balance.cw20() {
+        let cw20 = Cw20Contract(cw20_balance.address.clone());
+        let msg = cw20.call(Cw20ExecuteMsg::Transfer {
+            recipient: recepient,
+            amount: amount.unwrap_or(cw20_balance.amount),
+        })?;
+        return Ok(msg);
+    }
+    Err(ContractError::CouldNotCreateBankMessage {})
```

Comment:
code like this makes me nervous because even if it's correct now, it's not going to give us a compilation error if we extend this with new vesting types in the future. i feel `WrappedBalance` should be an enum.

to support native and cw20 types, and get compilation errors if new stuff get added, it could be:

```rust
pub enum WrappedBalance {
  denom: cw_denom::CheckedDenom,
  balance: Uint128,
}
```

using our [cw-denom](https://github.com/DA0-DA0/dao-contracts/blob/main/packages/cw-denom/src/lib.rs) package.

Reply:
Nice package, thank you!